aten/src/ATen/native/TensorShape.cpp

```diff
@@ -121,6 +121,93 @@ Tensor repeat(const Tensor& self, IntList repeats) {
   return result;
 }
 
+static std::vector<int64_t> infer_size(IntList shape, int64_t numel) {
+  auto res = shape.vec();
+  int64_t newsize = 1;
+  auto infer_dim = at::optional<int64_t>();
+  for (int64_t dim = 0, ndim = shape.size(); dim != ndim; dim++) {
+    if (shape[dim] == -1) {
+      if (infer_dim) {
+        throw std::runtime_error("only one dimension can be inferred");
+      }
+      infer_dim = dim;
+    } else if (shape[dim] >= 0) {
+      newsize *= shape[dim];
+    } else {
+      runtime_error("invalid shape dimension %zd", shape[dim]);
+    }
+  }
+
+  if (numel == newsize || (infer_dim && newsize > 0 && numel % newsize == 0)) {
+    if (infer_dim) {
+      res[*infer_dim] = numel / newsize;
+    }
+    if (numel == 0) {
+      // Collapse zero-element shapes into one dimension because TH handles zeros
+      // in sizes strangely: x.resize_(1, 0) has shape (1,). TODO: remove this
+      // once we have multi-dimensional empty tensors.
+      return {0};
+    }
+    return res;
+  }
+
+  std::ostringstream ss;
+  ss << "shape '" << shape << "' is invalid for input of size " << numel;
+  throw std::runtime_error(ss.str());
+}
+
+static at::optional<std::vector<int64_t>>
+compute_stride(const Tensor& self, IntList newshape) {
+  auto oldstride = self.strides();
+  auto oldshape = self.sizes();
+  if (oldshape.empty()) {
+    return std::vector<int64_t>(newshape.size(), 1);
+  }
+
+  std::vector<int64_t> newstride(newshape.size());
+  int64_t view_d = newshape.size() - 1;
+  // stride for each subspace in the chunk
+  int64_t chunk_base_stride = oldstride.back();
+  // numel in current chunk
+  int64_t tensor_numel = 1;
+  int64_t view_numel = 1;
+  for (int64_t tensor_d = oldshape.size() - 1; tensor_d >= 0; tensor_d--) {
+    tensor_numel *= oldshape[tensor_d];
+    // if end of tensor size chunk, check view
+    if ((tensor_d == 0) ||
+        (oldshape[tensor_d - 1] != 1 && oldstride[tensor_d - 1] != tensor_numel * chunk_base_stride)) {
+      while (view_d >= 0 && (view_numel < tensor_numel || newshape[view_d] == 1)) {
+        newstride[view_d] = view_numel * chunk_base_stride;
+        view_numel *= newshape[view_d];
+        view_d--;
+      }
+      if (view_numel != tensor_numel) {
+        return {};
```

Comment:
Is it possible to describe the conditions exactly *when* a reshape can be done as a view? That is to say, is there a simpler version of this algorithm which returns a boolean true/false whether or not reshape as a view can be done; simple enough that we can put it in the docs? How do we prove that this algorithm is tight (*if and only if* a view is possible, then this algorithm will find a view)?

Reply:
No and I don't think it is tight. It is a port of the existing view code to C++.

Reply:
To expand: I don't think it's important to document when a view is returned, because that strikes me as an implementation detail, which may change.

It would be nice if the algorithm were tight, but I don't think this implementation is and I don't think NumPy's implementation is either. I can't think of an efficient algorithm that's guaranteed to return always return a view if possible.

I'm not too concerned about this algorithm vs. other algorithms at this point. I think it's more important to match our implementation for `view`.

Reply:
On the other hand it is quite important to make the view/copy semantics clear. It makes reasoning about the program much easier (especially if you use in-place ops).

Reply:
Uhhh... if you need a copy use `clone()` if you need the same storage use `view()`. The semantics of `reshape()` are that it may or may not share the storage and you don't know beforehand. People seem to get along quite fine with that in NumPy.

Reply:
Yea, but NumPy makes doing in-place a bit awkward, so it's not used quite as often as in PyTorch

Reply:
@apaszke, I'm not sure what you're trying to get at:

1) That we should include a precise definition of when copying is performed in the docs?

The simplest definition is it returns a view when "subspaces of the strides are compatible" but this is neither precise nor particularly understandable. I'm skeptical that one can come up with a definition that's useful to readers.

2) That users often depend on the the shared storage semantics of view? 

This is (a) rare and (b) irrelevant because we still have `view()`. None of the calls to `view` in `examples`, `pyro`, or `OpenNMT` actually depend on shared storage semantics. The only place I know of in PyTorch core (outside of legacy nn) that depends on this is the numerical Jacobian computation.

The whole point of adding reshape is that the vast majority of call sites don't require shared storage semantics.

3) That people don't use mutating ops in NumPy?

`array[idx] = value` is pretty common (and of course +=, *=, etc.).

Reply:
I was just trying to say that it's good if we can have a simple explanation of share/copy semantics, but in this case I agree that it's better to just say "it can be a view or a copy, so you can't depend on this".

Reply:
Yeah, if that's the situation reshape is in, this is the comment I'm looking for. And maybe we can guarantee it's a view in simple situations, like when the input is contiguous.

Reply:
I've updated the documentation in _torch_docs.py to reflect this